scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py

```diff
@@ -0,0 +1,295 @@
+        self.presenter.add_group(group3)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Initialization
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def test_that_table_has_three_columns_when_initialized(self):
+        self.assertEqual(self.view.num_cols(), 3)
+
+    def test_that_model_is_initialized_as_empty(self):
+        self.assert_model_empty()
+
+    def test_that_view_is_initialized_as_empty(self):
+        self.assert_view_empty()
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Adding and removing groups
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def test_that_add_group_button_adds_group(self):
+        self.presenter.handle_add_group_button_clicked()
+        self.assertEqual(self.view.num_rows(), 1)
+        self.assertEqual(len(self.model.groups), 1)
+
+    def test_that_remove_group_button_removes_group(self):
+        self.add_three_groups_to_table()
+        self.presenter.handle_remove_group_button_clicked()
+        self.assertEqual(self.view.num_rows(), 2)
+
+    def test_that_add_group_button_adds_group_to_end_of_table(self):
+        self.add_three_groups_to_table()
+
+        self.presenter.add_group(MuonGroup(group_name="new", detector_ids=[4]))
+
+        self.assertEqual(self.view.get_table_item_text(self.view.num_rows() - 1, 0), "new")
+
+    def test_that_remove_group_button_removes_group_from_end_of_table(self):
+        self.add_three_groups_to_table()
+
+        self.presenter.handle_remove_group_button_clicked()
+
+        self.assertEqual(self.view.get_table_item_text(self.view.num_rows() - 1, 0), "my_group_1")
+
+    def test_that_highlighting_rows_and_clicking_remove_group_removes_the_selected_rows(self):
+        self.add_three_groups_to_table()
+        self.view._get_selected_row_indices = mock.Mock(return_value=[0, 2])
+
+        self.presenter.handle_remove_group_button_clicked()
+
+        self.assertEqual(self.view.get_table_item_text(0, 0), "my_group_1")
+        self.assertEqual(self.model.group_names, ["my_group_1"])
+
+    def test_that_cannot_add_more_than_20_rows(self):
+        for i in range(21):
+            self.presenter.handle_add_group_button_clicked()
+
+        self.assertEqual(self.view.num_rows(), 20)
+        self.assertEqual(len(self.model.groups), 20)
+
+    def test_that_trying_to_add_a_20th_row_gives_warning_message(self):
+        for i in range(21):
+            self.presenter.handle_add_group_button_clicked()
+
+        self.assertEqual(self.view.warning_popup.call_count, 1)
+
+    def test_that_remove_group_when_table_is_empty_does_not_throw(self):
+
+        self.presenter.handle_remove_group_button_clicked()
```

Comment:
should this have an assert throws nothing and a check that the warning_popup call count is 0?

Reply:
I think the test will automatically fail if an exception is thrown. I have added the check that the message box does not display.